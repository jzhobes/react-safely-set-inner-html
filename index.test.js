import React from 'react';
import renderer from 'react-test-renderer';
import ReactSafelySetInnerHTML from './index';

describe('ReactSafelySetInnerHTML', () => {
  it('renders simple text', () => {
    const component = renderer.create(
      <ReactSafelySetInnerHTML>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore
        magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
        consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
        pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est
        laborum.
      </ReactSafelySetInnerHTML>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('converts HTML attributes to React props and allows for the default tags', () => {
    const component = renderer.create(
      <ReactSafelySetInnerHTML>
        {`Lorem ipsum dolor
          <div class="div-class" style="border: none; max-font-size: 12; max-width: 100px;">div</div>
          <span class="span-class">span</span>
          <p class="p-class">p</p>
          <a class="link" href="https://www.google.com" rel="noreferrer" target="_blank">anchor</a>
          <b>bold</b>
          <i>italic</i>
          <strong>strong</strong>
          <small>small</small>
          <table>
            <caption>caption</caption>
            <colgroup>
              <col span="2">col</col>
            </colgroup>
            <thead>
              <tr>
                <th>th</th>
                <th>th</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>td</td>
                <td>td</td>
              </tr>
            </tbody>
            <tfoot></tfoot>
          </thead>
          <video controls> 
            <source src="video.webm" type="video/webm"> 
            <source src="video.ogv" type="video/ogg"> 
            <source src="video.mp4" type="video/mp4">
            <source src="video.3gp" type="video/3gp">
            Fallback text for video.
          </video>
          <audio controls>
            <source src="audio.ogg" type="audio/ogg">
            <source src="audio.mp3" type="audio/mpeg">
            Fallback text for audio.
          </audio>
          <iframe src="https://www.google.com">
        `}
      </ReactSafelySetInnerHTML>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('allows for all other tags if an excludedTags list is provided', () => {
    const component = renderer.create(
      <ReactSafelySetInnerHTML excludedTags={['form']}>
        {`
          <iframe allowfullscreen src="https://www.google.com"></iframe>
          <span class="span-class">span</span>
          <iframe allowfullscreen="false" src="https://www.microsoft.com"></iframe>
          <form target="target.php">
            <div>div</div>
            <button type="submit">submit</button>
          </form>
        `}
      </ReactSafelySetInnerHTML>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('allows for children as an array of strings', () => {
    const component = renderer.create(
      <ReactSafelySetInnerHTML excludedTags={['form']}>
        Lorem ipsum
        {'<span>span</span>'}
        {'<div>div</div>'}
      </ReactSafelySetInnerHTML>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('handles malformed HTML gracefully', () => {
    const component = renderer.create(
      <ReactSafelySetInnerHTML>
        Lorem ipsum
        {'<span   ><span><<<<asdf'}
        {'<span   ><s:::pan><<<<asdf'}
      </ReactSafelySetInnerHTML>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
