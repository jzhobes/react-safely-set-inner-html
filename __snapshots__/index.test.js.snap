// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactSafelySetInnerHTML allows for all other tags if an excludedTags list is provided 1`] = `
Array [
  " ",
  <iframe
    allowFullScreen={true}
    src="https://www.google.com"
  />,
  " ",
  <span
    className="span-class"
  >
    span
  </span>,
  " ",
  <iframe
    allowFullScreen={false}
    src="https://www.microsoft.com"
  />,
  " ",
  " ",
]
`;

exports[`ReactSafelySetInnerHTML allows for children as an array of strings 1`] = `
Array [
  "Lorem ipsum",
  <span>
    span
  </span>,
  <div>
    div
  </div>,
]
`;

exports[`ReactSafelySetInnerHTML converts HTML attributes to React props and allows for the default tags 1`] = `
Array [
  "Lorem ipsum dolor ",
  <div
    className="div-class"
    style={
      Object {
        "border": "none",
        "maxFontSize": "12",
        "maxWidth": "100px",
      }
    }
  >
    div
  </div>,
  " ",
  <span
    className="span-class"
  >
    span
  </span>,
  " ",
  <p
    className="p-class"
  >
    p
  </p>,
  " ",
  <a
    className="link"
    href="https://www.google.com"
    rel="noreferrer"
    target="_blank"
  >
    anchor
  </a>,
  " ",
  <b>
    bold
  </b>,
  " ",
  <i>
    italic
  </i>,
  " ",
  <strong>
    strong
  </strong>,
  " ",
  <small>
    small
  </small>,
  " ",
  <table>
     
    <caption>
      caption
    </caption>
     
    <colgroup>
       
      <col
        span={2}
      />
      col 
    </colgroup>
     
    <thead>
       
      <tr>
         
        <th>
          th
        </th>
         
        <th>
          th
        </th>
         
      </tr>
       
    </thead>
     
    <tbody>
       
      <tr>
         
        <td>
          td
        </td>
         
        <td>
          td
        </td>
         
      </tr>
       
    </tbody>
     
    <tfoot />
     
    <video
      controls={true}
    >
       
      <source
        src="video.webm"
        type="video/webm"
      />
       
      <source
        src="video.ogv"
        type="video/ogg"
      />
       
      <source
        src="video.mp4"
        type="video/mp4"
      />
       
      <source
        src="video.3gp"
        type="video/3gp"
      />
       Fallback text for video. 
    </video>
     
    <audio
      controls={true}
    >
       
      <source
        src="audio.ogg"
        type="audio/ogg"
      />
       
      <source
        src="audio.mp3"
        type="audio/mpeg"
      />
       Fallback text for audio. 
    </audio>
     
  </table>,
]
`;

exports[`ReactSafelySetInnerHTML handles malformed HTML gracefully 1`] = `
Array [
  "Lorem ipsum",
  <span>
    <span>
      &lt;&lt;&lt;
    </span>
  </span>,
  <span />,
]
`;

exports[`ReactSafelySetInnerHTML handles substitutions 1`] = `
Array [
  " ",
  <p>
     Hello, Bob 
  </p>,
  " ",
  <p>
     Lorem ipsum. 
  </p>,
  " ",
  <p>
     Lorem 
    <b>
      ipsum
    </b>
     dolor. 
  </p>,
  " ",
  <p>
     {noMatchFound} 
  </p>,
  " ",
]
`;

exports[`ReactSafelySetInnerHTML renders simple text 1`] = `"Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."`;
